src/System.Text.Primitives/System/Text/Encoders/Utf8Utility.cs

```diff
@@ -0,0 +1,357 @@
+
+namespace System.Buffers.Text
+{
+    /// <summary>
+    /// Contains utility methods for inspecting UTF-8 sequences.
+    /// </summary>
+    /// <remarks>
+    /// The UTF-8 specification can be found in the Unicode Standard, Sec. 3.9.
+    /// http://www.unicode.org/versions/Unicode10.0.0/ch03.pdf
+    /// </remarks>
+    public static class Utf8Utility
+    {
+        private static readonly byte[] _replacementCharAsUtf8 = new byte[] { 0xEF, 0xBF, 0xBD };
+
+        /// <summary>
+        /// The Unicode Replacement Character (U+FFFD) as the three-byte UTF-8 sequence [ EF BF BD ].
+        /// </summary>
+        public static ReadOnlySpan<byte> ReplacementCharacterByteSequence => _replacementCharAsUtf8;
+
+        /// <summary>
+        /// Given the first byte of a sequence, returns the expected number of continuation bytes
+        /// which should follow this byte.
+        /// </summary>
+        /// <remarks>
+        /// Returns 0 if <paramref name="firstByte"/> does not have any expected continuation bytes
+        /// or if <paramref name="firstByte"/> cannot begin a well-formed UTF-8 sequence.
+        /// </remarks>
+        public static int GetExpectedNumberOfContinuationBytes(byte firstByte)
+        {
+            if (firstByte < (byte)0xC2)
+            {
+                // ASII (one-byte sequence), or
+                // continuation byte (invalid as first byte, hence no expected followers), or
+                // [ C0..C1 ] (always invalid UTF-8 byte, hence no expected followers)
+                return 0;
+            }
+            else if (firstByte < (byte)0xE0)
+            {
+                // [ C2..DF ] can start a two-byte sequence
+                return 1;
+            }
+            else if (firstByte < (byte)0xF0)
+            {
+                // [ E0..EF ] can start a three-byte sequence
+                return 2;
+            }
+            else if (firstByte <= (byte)0xF4)
+            {
+                // [ F0..F4 ] can start a four-byte sequence
+                return 3;
+            }
+            else
+            {
+                // [ F5 .. FF ] (always invalid UTF-8 byte, hence no expected followers)
+                return 0;
+            }
+        }
+
+        /// <summary>
+        /// Returns the index of the first byte of the first invalid UTF-8 sequence in <paramref name="data"/>,
+        /// or -1 if <paramref name="data"/> is a well-formed UTF-8 string.
+        /// </summary>
+        public static int GetIndexOfFirstInvalidUtf8Sequence(ReadOnlySpan<byte> data) =>
+            Utf8Util.GetIndexOfFirstInvalidUtf8Sequence(data, out _, out _);
+
+        /// <summary>
+        /// Return <see langword="true"/> iff <paramref name="value"/> is an ASCII value (within the range 0-127, inclusive).
+        /// </summary>
+        public static bool IsAsciiValue(byte value) => (value < (byte)0x80);
```

Comment:
I have seen this method before. We shoudl have one in one place.

Reply:
Make sure these guys are annotated with AggresiveInline, there are edge cases where this kind of small methods are not inlined while when decorated most of the time they are honored.